### DISCUSSION TITLE
How can I pass parameters between refreshable functions?

### DISCUSSION THREAD

**Discussion by Aues6uen11Z** [zauberzeug/nicegui#2874] [Q&A] [answered]
### Question

Hi, I encountered a new problem while using NiceGUI. I have two functions, and the content of the second function depends on the first one. Therefore, I need to pass parameters between these two functions, and sometimes I also need to update their values simultaneously. Here is a simple example: 

```python
from nicegui import ui


@ui.refreshable
def sender():
    num = random.randint(1, 100)
    ui.label(f'I send {num}')
    return num

@ui.refreshable
def receiver(num):
    ui.label(f'I receive {num}')

def re():
    num = sender.refresh()
    receiver.refresh(num)

num = sender()
receiver(num)
ui.button(on_click=lambda: re())

ui.run()
```

The problem now is that the `refresh()` method cannot obtain the return value of the function, so it is difficult to communicate. Using global variables does not seem very elegant. Is there a better solution? Thanks for your help.

**Comment by falkoschindler** [comment]
Hi @Aues6uen11Z,
You're right, `refresh()` doesn't return anything at the moment. Instead it is designed to only consume data and generate UI from it. So I would restructure your code like this:
```py
def generate_data():
    return random.randint(1, 100)

@ui.refreshable
def receiver1(num):
    ui.label(f'Receiver 1: {num}')

@ui.refreshable
def receiver2(num):
    ui.label(f'Receiver 2: {num}')

def re():
    num = generate_data()
    receiver1.refresh(num)
    receiver2.refresh(num)

num = generate_data()
receiver1(num)
receiver2(num)
ui.button(on_click=re)
```

I'm not sure if this helps or if the refreshable UI function _really_ needs to create input for another refreshable.

**Reply by Aues6uen11Z** [reply]
Hi @falkoschindler  , thank you for your help. Unfortunately, this method is not very suitable for my situation. I want to generate an interface based on a json file, and when the user imports a new json configuration, I need to refresh the interface elements based on these configurations. For example:

```python
from nicegui import ui


json_config = ['tab1', 'tab2']  # extract from json file

@ui.refreshable
def sender():
    tab_list = []
    with ui.tabs() as drawer_tabs:
        for instance in json_config:
            tab = ui.tab(instance)
            tab_list.append(tab)
    return drawer_tabs, tab_list

@ui.refreshable
def receiver(drawer_tabs, tab_list):
    with ui.tab_panels(drawer_tabs, value=tab_list[0]):
        for tab in tab_list:
            with ui.tab_panel(tab):
                ui.label(f'other contents...')

def update():
    # Here I will add new tab in json file
    drawer_tabs, tab_list = sender.refresh()
    receiver.refresh(drawer_tabs, tab_list)

with ui.left_drawer():
    drawer_tabs, tab_list = sender()
receiver(drawer_tabs, tab_list)
ui.button(on_click=lambda: update())

ui.run()
```
The sender generates tabs according to the configuration file, and the receiver displays the corresponding tab_panel. However, since the sender is under the left_drawer, I can’t write these two functions as one, otherwise an warning will be reported: “Found top level layout element “LeftDrawer” inside element “RefreshableContainer”. Top level layout elements should not be nested but must be direct children of the page content. This will be raising an exception in NiceGUI 1.5.” Are there any other possible methods?

**Reply by Magic-wei** [reply, marked as answer]
Hi @Aues6uen11Z, I have tested your code and reproduced the issue. My suggestion is, what about define a class and share the data as class members? The variables are defined locally in the class, and you can just keep the code clean to maintain (since your app is just the class). I personally prefer to develop Python project this way and hope this can help. Here is an example I modified based on your code:

```python
from nicegui import ui


# demo class
class RefreshableTest():

    def __init__(self) -> None:
        # extract from json file
        self.json_config = ['tab1', 'tab2']

    @ui.refreshable
    def sender(self):
        tab_list = []
        with ui.tabs() as drawer_tabs:
            for instance in self.json_config:
                tab = ui.tab(instance)
                tab_list.append(tab)
        self.drawer_tabs = drawer_tabs
        self.tab_list = tab_list

    @ui.refreshable
    def receiver(self, drawer_tabs, tab_list):
        with ui.tab_panels(drawer_tabs, value=tab_list[0]):
            for tab in tab_list:
                with ui.tab_panel(tab):
                    ui.label(f'other contents...')

    def update(self):
        # Here the list is from inputs, you can change it to read from JSON file
        self.json_config = self.tabs_input.value.split(',')
        self.sender.refresh()
        self.receiver.refresh(self.drawer_tabs, self.tab_list)

    def run(self):
        with ui.left_drawer():
            self.sender()
            drawer_tabs = self.drawer_tabs
            tab_list = self.tab_list

        self.receiver(drawer_tabs, tab_list)

        self.tabs_input = ui.input('tabs:', value="1,2,tab3")
        ui.button(on_click=lambda: self.update())

        ui.run()


# run
app = RefreshableTest()
app.run()
```

**Reply by Aues6uen11Z** [reply]
Thanks @Magic-wei It seems like this is a good approach. I didn’t initially plan to use an object-oriented method, but now I may need to make significant changes to my code. Thank you very much.

edit: 才发现还是北理学长hhh，非常感谢
